src/common/mem.c

```diff
@@ -85,6 +87,166 @@ umempobj_settings_init(bool md_on_ssd)
+    -1, -1, 12, /* 512 bytes */
+    -1, 13,     /* 576 bytes (2.2 compatibility only) */
+    -1, -1, 14, /* 672 bytes (2.2 compatibility only) */
+    -1, -1, 15, /* 768 bytes (2.2 compatibility only) */
+};
+
+/** Create a slab within the pool.
+ *
+ *  \param	pool[IN]		Pointer to the persistent object.
+ *  \param	slab[IN]		Slab description
+ *
+ *  \return	zero on success, non-zero on failure.
+ */
+static int
+set_slab_desc(struct umem_pool *ph_p, struct umem_slab_desc *slab)
+{
+	PMEMobjpool			*pop;
+	struct pobj_alloc_class_desc	 pmemslab;
+	struct dav_alloc_class_desc	 davslab;
+	int				 rc = 0;
+
+	switch (daos_md_backend) {
+	static unsigned class_id = 10;
+
+	case DAOS_MD_PMEM:
+		pop = (PMEMobjpool *)ph_p->up_priv;
+
+		pmemslab.unit_size = slab->unit_size;
+		pmemslab.alignment = 0;
+		pmemslab.units_per_block = 1000;
+		pmemslab.header_type = POBJ_HEADER_NONE;
+		pmemslab.class_id = slab->class_id;
+		rc = pmemobj_ctl_set(pop, "heap.alloc_class.new.desc", &pmemslab);
+		/* update with the new slab id */
+		slab->class_id = pmemslab.class_id;
+		break;
+	case DAOS_MD_BMEM:
+		davslab.unit_size = slab->unit_size;
+		davslab.alignment = 0;
+		davslab.units_per_block = 1000;
+		davslab.header_type = DAV_HEADER_NONE;
+		davslab.class_id = slab->class_id;
+		rc = dav_class_register((dav_obj_t *)ph_p->up_priv, &davslab);
+		/* update with the new slab id */
+		slab->class_id = davslab.class_id;
+		break;
+	case DAOS_MD_ADMEM:
+		/* NOOP for ADMEM now */
+		slab->class_id = class_id++;
+		break;
+	default:
+		D_ASSERTF(0, "bad daos_md_backend %d\n", daos_md_backend);
+		break;
+	}
+	return rc;
+}
+
+static inline bool
+slab_registered(struct umem_pool *pool, unsigned int slab_id)
+{
+	D_ASSERT(slab_id < UMM_SLABS_CNT);
+	return pool->up_slabs[slab_id].class_id != 0;
+}
+
+static inline size_t
+slab_usize(struct umem_pool *pool, unsigned int slab_id)
+{
+	D_ASSERT(slab_id < UMM_SLABS_CNT);
+	return pool->up_slabs[slab_id].unit_size;
+}
+
+static inline uint64_t
+slab_flags(struct umem_pool *pool, unsigned int slab_id)
+{
+	D_ASSERT(slab_id < UMM_SLABS_CNT);
+	return (daos_md_backend == DAOS_MD_PMEM) ?
+		POBJ_CLASS_ID(pool->up_slabs[slab_id].class_id) :
+		DAV_CLASS_ID(pool->up_slabs[slab_id].class_id);
+}
+
+static inline void
+get_slab(struct umem_instance *umm, uint64_t *flags, size_t *size)
+{
+	struct umem_pool *pool;
+	size_t aligned_size = D_ALIGNUP(*size, 32);
+	int slab_idx;
+	int slab;
+
+	slab_idx = (aligned_size >> 5) - 1;
+	if (slab_idx >= ARRAY_SIZE(slab_map))
+		return;
+
+	if (slab_map[slab_idx] == -1) {
+		D_DEBUG(DB_MEM, "No slab %zu for allocation of size %zu, idx=%d\n", aligned_size,
+			*size, slab_idx);
+		return;
+	}
+
+	pool = umm->umm_pool;
+	D_ASSERT(pool != NULL);
+
+	slab = slab_map[slab_idx];
+
+	D_ASSERTF(!slab_registered(pool, slab) || aligned_size == slab_usize(pool, slab),
+		  "registered: %d, id: %d, size: %zu != %zu\n", slab_registered(pool, slab),
+		  slab, aligned_size, slab_usize(pool, slab));
```

Comment:
Minor: since now the slabs are registered on umem open, I think the slabs are always registered, so the assert could be changed to "(aligned_size == slab_usize(pool, slab))" only.

Reply:
I will address in future patch.